### PR TITLE
extract supplemental data elements from hqmf

### PR DESCRIPTION
**Pull request by lukoktonos**
This PR adds support for parsing Supplemental Data Elements, which will be used in Hybrid measures.

V3_IG_CQL_HQMF_R1_STU2_2017JUL_Vol1 Section 5.9: Supplemental Data Elements

Hybrid measure found at : https://gitlab.mitre.org/bonnie/internal-documentation/wikis/cql-measures-for-testing

In bonnie, load hybrid measure and do `
bonnie.mainView._view.measure.get('populations').first().get('supplemental_data_elements')` to see what is passed back to bonnie.

`extract_populations_and_criteria` was partially refactored in order to avoid too much duplication when modifying it for the child module HQMF2CQL.  I believe a full refactor would be too much effort for this task and have too many risks associated with it currently.

Pull requests into Health Data Standards require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.
 
**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/BONNIE-1443
- [x] Internal ticket links back to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] Code coverage has not gone down and all code touched or added is covered.
- [x] Bonnie bundler tests pass
- [x] Bonnie tests pass

Branch | Coverage
-- | --
master_bonnie |  6979 / 7703 LOC (90.6%) covered. <br> 2696 / 7801 LOC (34.56%) covered.
sde_parsing |  7016 / 7740 LOC (90.65%) covered.
 
Note: Coverage report was fixed in this PR so a single complete coverage report is now created.

**Cypress Reviewer:**
 
Name: @okeefm 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
 
**Bonnie Reviewer 1:**
 
Name: @holmesie 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

**Bonnie Reviewer 2:**
 
Name: @mayerm94 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

File: lib/hqmf-parser/2.0/document_helpers/doc_population_helper.rb [hunk 1]
```diff
@@ -50,6 +50,10 @@ def extract_populations_and_criteria
         population['title'] = title_def ? title_def.value : "Population #{population_index}"
 
         population['OBSERV'] = 'OBSERV' if has_observation
+
+        # Do additional processing for CQL
+        yield(population_def, population) if block_given?
```
Comment (holmesie) [inline lib/hqmf-parser/2.0/document_helpers/doc_population_helper.rb]: This shouldn't be done here, this is the base parser for QDM logic. The CQL specific parsing code is in `lib/hqmf-parser/cql/`. The CQL parsing code is essentially an extension to the 2.0 parser that handles the CQL specific stuff. If you need this functionality for CQL parsing, you can extend this class in that directory and implement the behavior there.

File: lib/hqmf-parser/cql/document_helpers/doc_population_helper.rb [hunk 2]
```diff
@@ -120,5 +122,20 @@ def extract_main_library
       end
     end
 
+    def extract_supplemental_data_elements(population_def, population)
+      begin
+        supplemental_data_elements_def = population_def.xpath("cda:component/cql-ext:supplementalDataElement")
```
Comment (holmesie) [inline lib/hqmf-parser/cql/document_helpers/doc_population_helper.rb]: Prefer single quotes for anything that isn't string interpolation.

File: lib/hqmf-parser/cql/document_helpers/doc_population_helper.rb [hunk 3]
```diff
@@ -120,5 +122,20 @@ def extract_main_library
...
+      rescue Nokogiri::XML::XPath::SyntaxError
```
Comment (holmesie) [inline lib/hqmf-parser/cql/document_helpers/doc_population_helper.rb]: This doesn't seem to me to be the ideal way of checking if that xpath is present. I'd look at nokogiri's `at` call (specifically `at_xpath`), which returns nil if there wasn't something (which you can test the truthiness of, rather than relying on a begin/rescue.

File: test/unit/hqmf/cql/hqmf_cql_test.rb [hunk 4]
```diff
@@ -31,5 +31,12 @@ def do_test(measure_filename, measure_name)
     diff = ''
     assert diff.empty?, 'Diff is not empty.'
   end
+  def test_supplemental_data_element_parsing
+
```
Comment (holmesie) [inline test/unit/hqmf/cql/hqmf_cql_test.rb]: I like this test! I'd just recommend removing this blank line.

File: lib/hqmf-parser/cql/document_helpers/doc_population_helper.rb [hunk 5]
```diff
@@ -120,5 +122,20 @@ def extract_main_library
...
+        # If the hqmf has no SDEs, it won't have the cql-ext namespace
+        return
+      end
+
+      supplemental_data_elements = []
+      supplemental_data_elements_def.each do |sde_def|
+        cql_definition_name = sde_def.at_xpath("cda:precondition/cda:criteriaReference/cda:id").attribute('extension').to_s.match(/"([^"]*)"/)[1]
```
Comment (holmesie) [inline lib/hqmf-parser/cql/document_helpers/doc_population_helper.rb]: Prefer single quotes for anything that isn't string interpolation.

**Review by holmesie — CHANGES_REQUESTED** [review]

**Comment by holmesie** [comment]
Also seeing this error:
```
HQMF2CQLTest
  test_2                                                          PASS (0.14s)
  test_supplemental_data_element_parsing                         ERROR (0.00s)
Errno::ENOENT:         Errno::ENOENT: No such file or directory @ rb_sysopen - test/fixtures/hqmf/cql/CCDELookback_v5_4_eCQM.xml
            /home/travis/build/projectcypress/health-data-standards/test/unit/hqmf/cql/hqmf_cql_test.rb:37:in `initialize'
            /home/travis/build/projectcypress/health-data-standards/test/unit/hqmf/cql/hqmf_cql_test.rb:37:in `open'
            /home/travis/build/projectcypress/health-data-standards/test/unit/hqmf/cql/hqmf_cql_test.rb:37:in `test_supplemental_data_element_parsing'
```

**Review by holmesie — DISMISSED** [review]

**Review by mayerm94 — DISMISSED** [review]
Changes make sense. Nice refactoring

**Review by holmesie — APPROVED** [review]
Re-review passes. Good catch on that extra `require`!

**Review by mayerm94 — APPROVED** [review]
New changes make sense. Simplecov fix works

**Review by okeefm — APPROVED** [review]
Looks good to me. Tested the `SimpleCov` fix, and it works (the difference in lines in `master_bonnie` was due to an `import` of some rake tasks in `Rakefile`).
